components/datetime/src/mock/mod.rs

```diff
@@ -51,9 +51,48 @@ pub fn parse_gregorian_from_str(input: &str) -> Result<DateTime<Gregorian>, Date
     let second: u8 = input[17..19].parse()?;
     #[allow(clippy::indexing_slicing)] // TODO(#1668) Clippy exceptions need docs or fixing.
     let fraction: u32 = if input.len() > 20 {
-        input[20..].parse().unwrap_or(0)
+        // Extract fractional input and trim any trailing zeros
+        let mut fraction = input[20..].trim_end_matches('0');
+        // Truncate to at most 9 digits
+        if fraction.len() > 9 {
+            fraction = &fraction[0..9];
+        }
+        let as_int: u32 = fraction.parse().unwrap_or(0);
+        // Convert fraction to nanoseconds
+        as_int * (10u32.pow(9 - fraction.len() as u32))
```

Comment:
Suggestion (optional; can do later if you like the idea): to avoid the `10u32.pow(...)`, you could have a stack-allocated fixed-length string of length 9, which is populated with `'0'` characters, then overwrite the first characters of that string with the characters from the input string, then parse the stack string, always resulting in a number of the correct magnitude

Reply:
Sure, I'll address this in my formatting patch.

Reply:
I started doing this, but this is in test code anyway, so I don't think it's worth making the change.